histogram/src/lib.rs

```diff
@@ -1,206 +1,121 @@
 
-    #[test]
-    fn percentiles_1() {
-        let histogram = Histogram::new(0, 2, 10).unwrap();
+    fn total_count(&self) -> u128;
 
-        for v in 1..1024 {
-            assert!(histogram.increment(v, 1).is_ok());
-            assert_eq!(histogram.percentile(0.0).map(|b| b.high()), Ok(1));
+    fn get_count(&self, index: usize) -> u64;
 
-            assert!(histogram.percentile(100.0).map(|b| b.high()).unwrap_or(0) >= v);
-            assert!(histogram.percentile(100.0).map(|b| b.low()).unwrap_or(0) <= v);
+    fn get_bucket(&self, index: usize) -> Option<Bucket> {
+        if index >= self.config().total_bins() {
+            return None;
         }
 
-        let percentiles: Vec<(u64, u64)> = histogram
-            .percentiles(&[1.0, 10.0, 25.0, 50.0, 75.0, 90.0, 99.0])
-            .unwrap()
-            .iter()
-            .map(|p| (p.bucket().low(), p.bucket().high()))
-            .collect();
-
-        // this histogram config doesn't have much resolution, which results in
-        // the upper percentiles falling into buckets that are rather wide
-        assert_eq!(
-            &percentiles,
-            &[
-                (8, 11),
-                (96, 127),
-                (256, 383),
-                (512, 767),
-                (768, 1023),
-                (768, 1023),
-                (768, 1023)
-            ]
-        );
+        Some(Bucket {
+            count: self.get_count(index),
+            lower: self.config().index_to_lower_bound(index),
+            upper: self.config().index_to_upper_bound(index),
+        })
     }
 
-    #[test]
-    fn percentiles_2() {
-        let histogram = Histogram::new(0, 5, 10).unwrap();
-
-        for v in 1..1024 {
-            assert!(histogram.increment(v, 1).is_ok());
-            assert_eq!(histogram.percentile(0.0).map(|b| b.high()), Ok(1));
+    fn percentiles(&self, percentiles: &[f64]) -> Result<Vec<(f64, Bucket)>, Error> {
+        // get the total count across all buckets
+        let total: u128 = self.total_count();
 
-            assert!(histogram.percentile(100.0).map(|b| b.high()).unwrap_or(0) >= v);
-            assert!(histogram.percentile(100.0).map(|b| b.low()).unwrap_or(0) <= v);
+        // if the histogram is empty, then we should return an error
+        if total == 0_u128 {
+            return Err(Error::Empty);
         }
 
-        let percentiles: Vec<(u64, u64)> = histogram
-            .percentiles(&[1.0, 10.0, 25.0, 50.0, 75.0, 90.0, 99.0])
-            .unwrap()
+        // sort the requested percentiles so we can find them in a single pass
+        let mut percentiles = percentiles.to_vec();
+        percentiles.sort_by(|a, b| a.partial_cmp(b).unwrap());
+
+        // get a sorted list of the counts needed for each percentile
+        let counts: Vec<u128> = percentiles
             .iter()
-            .map(|p| (p.bucket().low(), p.bucket().high()))
+            .map(|p| (p / 100.0 * total as f64).ceil() as u128)
```

Comment:
I think it might be possible. What we need for this to be an issue is for total count to be represented as a larger f64 and for p = 100.0. I think if total = 99_999_999 we would see this?

Changing it to be defensive seems like a good idea.

Reply:
Fixed. I think this can be resolved.